### ISSUE TITLE
Timing issues preventing Up Next working reliably

### ISSUE THREAD

**Issue by thebertster** [im85288/service.upnext#139]
I am continuing to have issues with Up Next working reliably on both my Kodi systems (one is a Raspberry Pi 3, the other a slightly faster Intel Celeron fanless PC both running LibreElec 9.2.0/Kodi Leia 18.5).

I thought I'd try to debug the behaviour with the current version of the service pulled directly from the repo (last commit being f18864b on December 12).

There seems to be a timing issue with api.play_kodi_item() being invoked at the same time as the previous episode's playback is in the process of stopping. This appears to cause the next episode not to play - on my systems, this happens almost every time (maybe about 10% of the time the next episode does play).

The debug logs when this happens are as below. Playback of the previous episode has finished as can be seen by the VideoPlayer::OnExit() event being logged. You can then see the debug message "playing media episode" logged and then the VideoPlayer::OpenFile call is made. Playback of the next episode does not start - the VideoPlayer just exits.. The next episode never ends up playing successfully - I guess this is a Kodi VideoPlayer "issue" in that sense, but realistically it is only with something like Up Next that the precise timing needed to trigger the behaviour would be seen.

```
2019-12-16 23:35:54.556 T:140524635789056  NOTICE: [service.upnext] UpNextMonitor -> Show notification as episode (of length 1269 secs) ends in 15 secs
2019-12-16 23:35:54.559 T:140524635789056  NOTICE: [service.upnext] Api -> Got active player {u'jsonrpc': u'2.0', u'id': 1, u'result': [{u'playerid': 1, u'type': u'video', u'playertype': u'internal'}]}
2019-12-16 23:35:54.561 T:140524635789056  NOTICE: [service.upnext] Api -> Getting details of now playing media
2019-12-16 23:35:54.611 T:140524635789056  NOTICE: [service.upnext] Api -> Got details of now playing media {...removed JSON!...}
2019-12-16 23:35:55.711 T:140524635789056  NOTICE: [service.upnext] Api -> Got details of next up episode {...removed JSON!...}
2019-12-16 23:35:55.823 T:140524635789056  NOTICE: [service.upnext] PlaybackManager -> episode details {...removed JSON!...}
2019-12-16 23:35:55.829 T:140524635789056  NOTICE: [service.upnext] PlaybackManager -> played in a row settings 3
2019-12-16 23:35:55.831 T:140524635789056  NOTICE: [service.upnext] PlaybackManager -> played in a row 1
2019-12-16 23:35:55.832 T:140524635789056  NOTICE: [service.upnext] PlaybackManager -> showing next up page as played in a row is 1
2019-12-16 23:36:08.328 T:140523134215936  NOTICE: CVideoPlayer::OnExit()
2019-12-16 23:36:08.328 T:140523134215936  NOTICE: VideoPlayer: eof, waiting for queues to empty
2019-12-16 23:36:08.329 T:140523134215936  NOTICE: Closing stream player 1
2019-12-16 23:36:08.329 T:140523134215936  NOTICE: CDVDMessageQueue(audio)::WaitUntilEmpty
2019-12-16 23:36:08.353 T:140523134215936  NOTICE: Waiting for audio thread to exit
2019-12-16 23:36:08.354 T:140524111869696   ERROR: Got MSGQ_ABORT or MSGO_IS_ERROR return true
2019-12-16 23:36:08.354 T:140524111869696  NOTICE: thread end: CVideoPlayerAudio::OnExit()
2019-12-16 23:36:08.354 T:140523134215936  NOTICE: Closing audio device
2019-12-16 23:36:08.419 T:140524635789056  NOTICE: [service.upnext] PlaybackManager -> playing media episode
2019-12-16 23:36:08.462 T:140524635789056  NOTICE: [service.upnext] UpNextMonitor -> Up Next style autoplay succeeded
2019-12-16 23:36:08.532 T:140525870988928  NOTICE: VideoPlayer::OpenFile: smb://manasa/TV/The Big Bang Theory/Season 1/The Big Bang Theory - 1x03 - The Fuzzy Boots Corollary.mp4
2019-12-16 23:36:08.754 T:140523134215936  NOTICE: Deleting audio codec
2019-12-16 23:36:08.755 T:140523134215936  NOTICE: Closing stream player 2
2019-12-16 23:36:08.755 T:140523134215936  NOTICE: CDVDMessageQueue(video)::WaitUntilEmpty
2019-12-16 23:36:08.755 T:140523134215936  NOTICE: waiting for video thread to exit
2019-12-16 23:36:08.755 T:140523846481664   ERROR: Got MSGQ_ABORT or MSGO_IS_ERROR return true
2019-12-16 23:36:08.755 T:140523846481664  NOTICE: thread end: video_thread
2019-12-16 23:36:08.755 T:140523134215936  NOTICE: deleting video codec
2019-12-16 23:36:08.834 T:140525870988928  NOTICE: CVideoPlayer::CloseFile()
2019-12-16 23:36:08.834 T:140525870988928  NOTICE: VAAPI::Close
2019-12-16 23:36:08.834 T:140523872446208  NOTICE: COutput::OnExit: Output Thread terminated
2019-12-16 23:36:08.834 T:140525870988928  NOTICE: VAAPI::Close - closing decoder context
2019-12-16 23:36:08.837 T:140525870988928  NOTICE: VideoPlayer: waiting for threads to exit
2019-12-16 23:36:08.837 T:140525870988928  NOTICE: VideoPlayer: finished waiting
2019-12-16 23:36:08.837 T:140525870988928  NOTICE: CVideoPlayer::CloseFile()
2019-12-16 23:36:08.838 T:140525870988928  NOTICE: VideoPlayer: waiting for threads to exit
2019-12-16 23:36:08.838 T:140525870988928  NOTICE: VideoPlayer: finished waiting
2019-12-16 23:36:08.844 T:140525870988928  NOTICE: CWinSystemX11::SetFullScreen - calling xrand
```

If I change the timing when the playback of the next episode is triggered, e.g. by starting the process just one second earlier (i.e. just before the previous episode has finished playing) by changing:

```
while (self.player.isPlaying() and (total_time - play_time > 2)
```

then the problem does not seem to occur - the next episode plays reliably - with the obvious cosmetic issue that the countdown timer only gets down to 2!

```
2019-12-16 23:53:23.775 T:139661750015744  NOTICE: [service.upnext] PlaybackManager -> played in a row settings 3
2019-12-16 23:53:23.776 T:139661750015744  NOTICE: [service.upnext] PlaybackManager -> played in a row 1
2019-12-16 23:53:23.778 T:139661750015744  NOTICE: [service.upnext] PlaybackManager -> showing next up page as played in a row is 1
2019-12-16 23:53:35.255 T:139661750015744  NOTICE: [service.upnext] PlaybackManager -> playing media episode
2019-12-16 23:53:35.294 T:139661750015744  NOTICE: [service.upnext] UpNextMonitor -> Up Next style autoplay succeeded
2019-12-16 23:53:35.324 T:139663115423360  NOTICE: VideoPlayer::OpenFile: smb://manasa/TV/The Big Bang Theory/Season 1/The Big Bang     Theory - 1x03 - The Fuzzy Boots Corollary.mp4
2019-12-16 23:53:35.462 T:139660393379584  NOTICE: Creating InputStream
2019-12-16 23:53:35.503 T:139660393379584  NOTICE: Creating Demuxer
2019-12-16 23:53:35.582 T:139660393379584  NOTICE: Opening stream: 0 source: 256
```

I also looked at instead adding a delay before playing the next video to give time for the previous episode's player to fully stop. While this does mean that playback of the next episode starts successfully, it causes a secondary issue in that the onPlaybackEnded() event will fire for the previous episode, causing State() to be reset which in turn prevents still watching tracking from working. I guess this is a race condition that could also potentially happen in other situations, perhaps on a very fast system where playback of the previous episode ends and onPlaybackEnded() fires before the next episode's playback is started...

**Comment by dagwieers** [comment]
So we have been looking at improvements to how Kodi's Player behaves when transitioning from current to next episode, and one of the ideas is to add the next episode to the playlist and then starting the next item on the playlist. Overall this seems to offer the best experience.

The main questions is at what time do we update the playlist when the transition is an automatic transition. Doing this too early (on notification time) the user may Cancel and we have a modified playlist. Doing this too late will fail to play the next episode.

PS In your description you start saying that it works reliably on RPi3 and Intel Celeron. So it is unclear to me when/where it fails 90% of the time.

**Comment by thebertster** [comment]
Badly worded opening sentence. I have issues with reliability on both of those systems - 90% of the time on both systems, the next video does not play.

**Comment by dagwieers** [comment]
I have Raspberry Pi 3B (and 3B+) and 4, all running LibreELEC 9.2.0 and they all appear to work fine.

One of the suspected causes could be that too many add-ons are running services with event handlers for Player or Monitor. So when an event occurs (i.e. onPlayBackEnder or onPlayBackStopped) they are all triggered and handling these events causing such timing issues.

In Kodi you can go to **Settings » System » Add-ons » Running** to see which and how many add-ons are running services in the background. And you could try to disable some or all of them to see if this affects how Kodi behaves.

IIRC the services I have running are Up Next, Youtube, Netflix, VRT NU and VTM GO. Nothing else.

**Comment by thebertster** [comment]
I had three other services which subscribed to notifications (User Rating, Skin Helper Service and Skin Helper Widgets - the latter two being dependencies for many skins, including the Titan BINGIE skin I use). I uninstalled User Rating, switching back to Estuary and disabled the two Skin Helper services (rebooted just to be sure no remnants were left running), but this made no difference - out of 10 attempts, only 1 successfully played the next episode; same symptoms in the logs.

The only other thing I can think of is that all my media is streamed from a NAS using SMB (Ethernet on the HTPC, WiFi on the Pi). Possibly it takes the video player a little longer to close the playback stream compared with a local file - from the logs, there is typically around 400-500ms (it is a little longer on the Pi) between the CVideoPlayer::OnExit() log entry and the CVideoPlayer::CloseFile(), with the log for Up Next requesting playback of the next episode occurring between these two events.

If I get a chance, I will try copying some episodes locally to the device. However, starting the playback just one second earlier as per the previous post seems to be enough to work around this timing issue consistently on both systems.

**Comment by dagwieers** [comment]
@thebertster Thanks for ruling out that hypothesis. The use of SMB I considered as well, but looked less likely. I am using NFS myself because it appeared to work better than SMB (however this was before I was using Up Next) however timing issues could also be related to specific network/NAS performance or issues. Up Next should not be sensitive to this, so it is a bug if this is the cause. (Would be nice to pin it on something though)

Starting playback a second earlier looks more like a workaround than a proper fix. It shouldn't get down to timing issues like this, but I don't know what the proper fix ought to be, yet.

**Comment by AloofSage** [comment]
I have this same issue. My log looks very similar to what @thebertster posted. I use Emby server to manage my media and so I tried disabling the Kodi addon for that to see if it would help but it didn't. It happens in both the Aura skin which i normally use and Estuary.

**Comment by dagwieers** [comment]
Verify that the master branch fixes your issue and please report back !

**Comment by thebertster** [comment]
Looking good - been running with the new build for a couple of days and it seems to be working 100% now!

**Comment by thebertster** [comment]
@dagwieers - One thing I have noticed with the master build is that the still watching functionality doesn't seem to be working. I never see the still watching popup and the next episode just automatically plays every time.

**Comment by dagwieers** [comment]
@thebertster Something like this was reported before in #135, but closed because of lack of information. This one is not that easy to troubleshoot. I should probably look at improving our testing framework to mimic this behaviour and allows for easier testing.

**Comment by thebertster** [comment]
@dagwieers - I've debugged what's going on and it seems to me there are multiple fundamental issues with the new mechanism to queue the next item using playlists.

The cause of the "still watching" dialog not displaying is simply that each time the next episode is played, it is as the result of the previous episode having played to the end (triggering the queued episode in the playlist). However this means the onPlayBackEnded() event will fire, which resets the state. We need to not reset state in the case where playback ended while the next episode was queued up.

However there are other issues. Firstly, now we are queuing the next episode to play but there is no mechanism currently to "dequeue" the next episode if the user cancels the playback of the next episode - it will play anyway. We need to "dequeue" in the case the user cancels playback of the next episode.

Also, in the case of displaying the still watching dialog, we have already queued the next episode to play, so regardless of what the user does with the still watching dialog, the next episode is going to play regardless. In this case, we should only queue the next episode if the user has explicitly acknowledged they are still watching.

I've forked the repo and have fixed most of the above, but am currently having issues with reliably dequeuing the next episode in the event of the user cancelling. I will work further on this if you like and submit a PR if I can get it working reliably.
